File: temporal-research/07_provisioning/workflows.py
import asyncio
from datetime import timedelta
from typing import Optional

from temporalio import workflow
from temporalio.exceptions import ApplicationError

with workflow.unsafe.imports_passed_through():
    from activities import (
        provision_cache,
        provision_database,
        provision_load_balancer,
        provision_server,
        provision_subnet,
        provision_vpc,
        register_dns_records,
        run_health_check,
    )
    from models import (
        CacheResult,
        DatabaseResult,
        DnsRecord,
        EnvironmentManifest,
        LoadBalancerResult,
        ProvisioningRequest,
        ServerResult,
        SubnetResult,
        VpcResult,
    )

_ACT = {"start_to_close_timeout": timedelta(seconds=60)}


# ── Child workflow ─────────────────────────────────────────────────────────────

@workflow.defn
class ServerProvisioningWorkflow:
    """
    Provisions a single server. Used as a child workflow so each server
    gets its own workflow ID, event history, and retry scope.

    In a real system this workflow would be more complex: install software,
    register with a service mesh, run smoke tests, etc.
    """

    @workflow.run
    async def run(
        self,
        vpc_id: str,
        subnet_id: str,
        instance_type: str,
        az: str,
        environment: str,
    ) -> ServerResult:
        return await workflow.execute_activity(
            provision_server,
            args=[vpc_id, subnet_id, instance_type, az, environment],
            **_ACT,
        )


# ── Main workflow ──────────────────────────────────────────────────────────────

@workflow.defn
class EnvironmentProvisioningWorkflow:
    """
    Provisions a complete cloud environment in five phases.

    Phase 1 — Network (sequential then parallel)
      Provision VPC first, then all subnets concurrently across 3 AZs.

    Phase 2 — Compute + Data (fully parallel, some conditional)
      Fan out N server child-workflows concurrently.
      Simultaneously provision database (if requested) and cache (if requested).
      All three groups run in parallel — servers don't wait for DB and vice versa.

    Phase 3 — Load Balancer (conditional)
      Only provisioned when server_count > 1. Skipped for single-server envs.

    Phase 4 — DNS
      Build records for every endpoint, register them all concurrently.

    Phase 5 — Health Checks (parallel)
      Probe every provisioned resource simultaneously.

    Signals: cancel_provisioning — soft-cancel between phases
    Queries: get_progress        — live phase + resource inventory
    """

    def __init__(self) -> None:
        self._phase = "pending"
        self._progress: dict = {}
        self._cancel_requested = False

    @workflow.signal
    async def cancel_provisioning(self) -> None:
        workflow.logger.info("Cancellation requested")
        self._cancel_requested = True

    @workflow.query
    def get_progress(self) -> dict:
        return {"phase": self._phase, **self._progress}

    @workflow.run
    async def run(self, req: ProvisioningRequest) -> EnvironmentManifest:
        wf_id = workflow.info().workflow_id

        # ── Phase 1: Network ──────────────────────────────────────────────────
        self._phase = "network"

        vpc: VpcResult = await workflow.execute_activity(
            provision_vpc, req.region, **_ACT
        )
        self._progress["vpc_id"] = vpc.vpc_id

        # Subnets across 3 AZs — provision all three simultaneously
        azs = [f"{req.region}a", f"{req.region}b", f"{req.region}c"]
        subnets: list[SubnetResult] = await asyncio.gather(*[
            workflow.execute_activity(
                provision_subnet,
                args=[vpc.vpc_id, az, f"10.0.{i}.0/24"],
                **_ACT,
            )
            for i, az in enumerate(azs)
        ])
        self._progress["subnet_ids"] = [s.subnet_id for s in subnets]

        self._check_cancel("after network phase")

        # ── Phase 2: Compute + Database + Cache (all parallel) ────────────────
        self._phase = "compute"

        # N server child-workflows, round-robining across subnets/AZs
        server_coros = [
            workflow.execute_child_workflow(
                ServerProvisioningWorkflow.run,
                args=[
                    vpc.vpc_id,
                    subnets[i % len(subnets)].subnet_id,
                    req.instance_type,
                    azs[i % len(azs)],
                    req.environment,
                ],
                id=f"{wf_id}-server-{i}",
            )
            for i in range(req.server_count)
        ]

        # Conditional coroutines — resolve to None when not requested
        async def maybe_db() -> Optional[DatabaseResult]:
            if not req.include_database:
                return None
            return await workflow.execute_activity(
                provision_database,
                args=[vpc.vpc_id, [s.subnet_id for s in subnets],
                      req.db_instance_type, req.environment],
                **_ACT,
            )

        async def maybe_cache() -> Optional[CacheResult]:
            if not req.include_cache:
                return None
            return await workflow.execute_activity(
                provision_cache,
                args=[vpc.vpc_id, subnets[0].subnet_id, req.environment],
                **_ACT,
            )

        # asyncio.gather runs all concurrently:
        #   - inner gather fans out N server child-workflows
        #   - maybe_db() and maybe_cache() run in parallel with the servers
        servers, database, cache = await asyncio.gather(
            asyncio.gather(*server_coros),
            maybe_db(),
            maybe_cache(),
        )

        self._progress["server_ids"] = [s.server_id for s in servers]
        if database:
            self._progress["database"] = database.endpoint
        if cache:
            self._progress["cache"] = cache.endpoint

        self._check_cancel("after compute phase")

        # ── Phase 3: Load Balancer (conditional) ─────────────────────────────
        self._phase = "load_balancer"
        load_balancer: Optional[LoadBalancerResult] = None

        if req.server_count > 1:
            load_balancer = await workflow.execute_activity(
                provision_load_balancer,
                args=[
                    vpc.vpc_id,
                    [s.subnet_id for s in subnets],
                    [s.server_id for s in servers],
                    req.environment,
                ],
                **_ACT,
            )
            self._progress["load_balancer"] = load_balancer.dns_name
        else:
            workflow.logger.info("Skipping load balancer (single server)")

        # ── Phase 4: DNS ──────────────────────────────────────────────────────
        self._phase = "dns"

        dns_records = _build_dns_records(req.environment, servers, load_balancer, database, cache)
        registered = await workflow.execute_activity(
            register_dns_records, args=[dns_records, req.environment], **_ACT
        )
        self._progress["dns_records"] = len(registered)

        # ── Phase 5: Health Checks (all parallel) ────────────────────────────
        self._phase = "health_checks"

        check_targets = [(s.server_id, s.private_ip) for s in servers]
        if load_balancer:
            check_targets.append((load_balancer.lb_id, load_balancer.dns_name))
        if database:
            check_targets.append((database.db_id, database.endpoint))
        if cache:
            check_targets.append((cache.cache_id, cache.endpoint))

        health_results = await asyncio.gather(*[
            workflow.execute_activity(
                run_health_check, args=[rid, ep], **_ACT
            )
            for rid, ep in check_targets
        ])

        healthy = sum(1 for r in health_results if r["healthy"])
        self._progress["health"] = f"{healthy}/{len(health_results)} healthy"

        self._phase = "complete"

        return EnvironmentManifest(
            environment=req.environment,
            region=req.region,
            vpc=vpc,
            subnets=subnets,
            servers=list(servers),
            database=database,
            cache=cache,
            load_balancer=load_balancer,
            dns_records=registered,
            status="ready" if healthy == len(health_results) else "degraded",
        )

    def _check_cancel(self, checkpoint: str) -> None:
        if self._cancel_requested:
            raise ApplicationError(
                f"Provisioning cancelled at {checkpoint}",
                non_retryable=True,
            )


# ── Helpers ────────────────────────────────────────────────────────────────────

def _build_dns_records(
    env: str,
    servers: list[ServerResult],
    lb: Optional[LoadBalancerResult],
    db: Optional[DatabaseResult],
    cache: Optional[CacheResult],
) -> list[DnsRecord]:
    records: list[DnsRecord] = []

    entry_point = lb.dns_name if lb else servers[0].private_ip
    records.append(DnsRecord(name=f"{env}.internal", value=entry_point))

    for i, s in enumerate(servers):
        records.append(DnsRecord(name=f"server-{i}.{env}.internal", value=s.private_ip))

    if db:
        records.append(DnsRecord(name=f"db.{env}.internal", value=db.endpoint, record_type="CNAME"))

    if cache:
        records.append(DnsRecord(name=f"cache.{env}.internal", value=cache.endpoint, record_type="CNAME"))

    return records

File: temporal-research-go/07_provisioning/workflows.go
package provisioning

import (
	"fmt"
	"time"

	"go.temporal.io/sdk/workflow"
)

// ServerWorkflow is a child workflow that provisions a single server.
func ServerWorkflow(ctx workflow.Context, subnetID, az string, idx int) (ServerResult, error) {
	ao := workflow.ActivityOptions{
		StartToCloseTimeout: 30 * time.Second,
	}
	actCtx := workflow.WithActivityOptions(ctx, ao)

	var result ServerResult
	err := workflow.ExecuteActivity(actCtx, ProvisionServer, subnetID, az, idx).Get(actCtx, &result)
	if err != nil {
		return ServerResult{}, fmt.Errorf("ProvisionServer failed (idx=%d): %w", idx, err)
	}
	return result, nil
}

// serverChildResult carries the result (or error) from a parallel server child workflow.
type serverChildResult struct {
	result ServerResult
	err    error
}

// subnetResult carries a subnet result from a parallel provisioning goroutine.
type subnetResult struct {
	result SubnetResult
	err    error
}

// EnvironmentWorkflow orchestrates the full multi-phase provisioning of an environment.
func EnvironmentWorkflow(ctx workflow.Context, req ProvisionRequest) (EnvManifest, error) {
	logger := workflow.GetLogger(ctx)

	ao := workflow.ActivityOptions{
		StartToCloseTimeout: 60 * time.Second,
	}
	actCtx := workflow.WithActivityOptions(ctx, ao)

	// Phase 1: Provision VPC
	var vpc VPCResult
	if err := workflow.ExecuteActivity(actCtx, ProvisionVPC, req).Get(actCtx, &vpc); err != nil {
		return EnvManifest{}, fmt.Errorf("ProvisionVPC failed: %w", err)
	}
	logger.Info("VPC provisioned", "vpcID", vpc.VPCID, "cidr", vpc.CIDR)

	// Phase 2: Provision 3 subnets in parallel
	azs := []string{"us-east-1a", "us-east-1b", "us-east-1c"}
	subnetCh := workflow.NewBufferedChannel(ctx, len(azs))

	for _, az := range azs {
		az := az
		workflow.Go(ctx, func(gCtx workflow.Context) {
			var sr SubnetResult
			err := workflow.ExecuteActivity(actCtx, ProvisionSubnet, vpc.VPCID, az).Get(actCtx, &sr)
			subnetCh.Send(gCtx, subnetResult{result: sr, err: err})
		})
	}

	subnets := make([]SubnetResult, 0, len(azs))
	for i := 0; i < len(azs); i++ {
		var sr subnetResult
		subnetCh.Receive(ctx, &sr)
		if sr.err != nil {
			return EnvManifest{}, fmt.Errorf("ProvisionSubnet failed: %w", sr.err)
		}
		subnets = append(subnets, sr.result)
		logger.Info("Subnet provisioned", "subnetID", sr.result.SubnetID, "az", sr.result.AZ)
	}

	// Phase 3: Servers + optional DB + optional Cache in parallel
	optCount := 0
	if req.IncludeDB {
		optCount++
	}
	if req.IncludeCache {
		optCount++
	}

	serverCh := workflow.NewBufferedChannel(ctx, req.ServerCount)

	type optionalResult struct {
		db    *DatabaseResult
		cache *CacheResult
		err   error
	}
	optCh := workflow.NewBufferedChannel(ctx, optCount+1) // +1 to avoid zero-size

	// Launch server child workflows.
	for i := 0; i < req.ServerCount; i++ {
		idx := i
		subnet := subnets[idx%len(subnets)]
		workflow.Go(ctx, func(gCtx workflow.Context) {
			cwo := workflow.ChildWorkflowOptions{
				WorkflowExecutionTimeout: 60 * time.Second,
			}
			childCtx := workflow.WithChildOptions(gCtx, cwo)
			var sr ServerResult
			err := workflow.ExecuteChildWorkflow(childCtx, ServerWorkflow, subnet.SubnetID, subnet.AZ, idx+1).Get(childCtx, &sr)
			serverCh.Send(gCtx, serverChildResult{result: sr, err: err})
		})
	}

	// Optionally launch database provisioning.
	if req.IncludeDB {
		workflow.Go(ctx, func(gCtx workflow.Context) {
			var dr DatabaseResult
			err := workflow.ExecuteActivity(actCtx, ProvisionDatabase, vpc.VPCID).Get(actCtx, &dr)
			if err != nil {
				optCh.Send(gCtx, optionalResult{err: fmt.Errorf("ProvisionDatabase: %w", err)})
				return
			}
			optCh.Send(gCtx, optionalResult{db: &dr})
		})
	}

	// Optionally launch cache provisioning.
	if req.IncludeCache {
		workflow.Go(ctx, func(gCtx workflow.Context) {
			var cr CacheResult
			err := workflow.ExecuteActivity(actCtx, ProvisionCache, vpc.VPCID).Get(actCtx, &cr)
			if err != nil {
				optCh.Send(gCtx, optionalResult{err: fmt.Errorf("ProvisionCache: %w", err)})
				return
			}
			optCh.Send(gCtx, optionalResult{cache: &cr})
		})
	}

	// Collect server results.
	servers := make([]ServerResult, 0, req.ServerCount)
	for i := 0; i < req.ServerCount; i++ {
		var sr serverChildResult
		serverCh.Receive(ctx, &sr)
		if sr.err != nil {
			return EnvManifest{}, fmt.Errorf("ServerWorkflow failed: %w", sr.err)
		}
		servers = append(servers, sr.result)
		logger.Info("Server provisioned", "serverID", sr.result.ServerID, "ip", sr.result.IP)
	}

	// Collect optional results.
	var dbResult *DatabaseResult
	var cacheResult *CacheResult
	for i := 0; i < optCount; i++ {
		var or optionalResult
		optCh.Receive(ctx, &or)
		if or.err != nil {
			return EnvManifest{}, or.err
		}
		if or.db != nil {
			dbResult = or.db
			logger.Info("Database provisioned", "endpoint", or.db.Endpoint)
		}
		if or.cache != nil {
			cacheResult = or.cache
			logger.Info("Cache provisioned", "endpoint", or.cache.Endpoint)
		}
	}

	// Phase 4: Optional Load Balancer
	var lbResult *LBResult
	if req.IncludeLB {
		serverIDs := make([]string, len(servers))
		for i, s := range servers {
			serverIDs[i] = s.ServerID
		}
		var lb LBResult
		if err := workflow.ExecuteActivity(actCtx, ProvisionLoadBalancer, vpc.VPCID, serverIDs).Get(actCtx, &lb); err != nil {
			return EnvManifest{}, fmt.Errorf("ProvisionLoadBalancer failed: %w", err)
		}
		lbResult = &lb
		logger.Info("Load balancer provisioned", "dns", lb.DNS)
	}

	// Phase 5: DNS registration + parallel health checks
	var dnsRecord *DNSRecord
	if lbResult != nil {
		var dns DNSRecord
		if err := workflow.ExecuteActivity(actCtx, RegisterDNS, req.Env, lbResult.DNS).Get(actCtx, &dns); err != nil {
			return EnvManifest{}, fmt.Errorf("RegisterDNS failed: %w", err)
		}
		dnsRecord = &dns
		logger.Info("DNS registered", "name", dns.Name, "target", dns.Target)
	}

	// Health-check each server in parallel.
	hcCh := workflow.NewBufferedChannel(ctx, len(servers))
	for _, srv := range servers {
		ip := srv.IP
		workflow.Go(ctx, func(gCtx workflow.Context) {
			var ok bool
			err := workflow.ExecuteActivity(actCtx, RunHealthCheck, ip).Get(actCtx, &ok)
			if err != nil {
				logger.Info("Health check failed", "ip", ip, "error", err)
			} else {
				logger.Info("Health check passed", "ip", ip, "ok", ok)
			}
			hcCh.Send(gCtx, struct{}{})
		})
	}
	for i := 0; i < len(servers); i++ {
		hcCh.Receive(ctx, nil)
	}

	return EnvManifest{
		Env:     req.Env,
		VPC:     vpc,
		Subnets: subnets,
		Servers: servers,
		DB:      dbResult,
		Cache:   cacheResult,
		LB:      lbResult,
		DNS:     dnsRecord,
	}, nil
}
